lib/telemetry_registry.ex

```diff
@@ -0,0 +1,253 @@
+
+  @typedoc """
+  A string representation of the measurements emitted with the event. This should resemble a typespec but is
+  not limited to the typespec format, i.e. you can include a comment on which unit a value is in. The objective
+  is to inform users.
+  """
+  @type event_measurements() :: :telemetry_registry.event_measurements()
+
+  @typedoc """
+  A string representation of the metadata emitted with the event. This should resemble a typespec but is
+  not limited to the typespec format, i.e. you can include comments or example values. The objective
+  is to inform users what is available.
+  """
+  @type event_metadata() :: :telemetry_registry.event_metadata()
+
+  @typedoc """
+  A map of event definition meta for an event containing the event, measurements, and metadata descriptions
+  if the event was declared with an event definition. Otherwise, this value will be an empty map.
+  """
+  @type event_meta() :: :telemetry_registry.event_meta()
+
+  @typedoc """
+  A list of spannable events known to the registry in the format of `{event_prefix, event_suffixes}`. For
+  example, given events `[:my_app, :request, :start], [:my_app, :request, :stop], [:my_app, :request, :exception]`
+  a span can be created from the `:start` -> `:stop` or the `:start` -> `:exception` events. These are aggregated
+  as a spannable event `{[:my_app, :request], [:start, :stop, :exception]}`.
+  """
+  @type spannable_event() :: :telemetry_registry.spannable_event()
+
+
+  defmacro __using__(_opts) do
+    quote do
+      import unquote(__MODULE__), only: [format_telemetry_docs: 1, telemetry_event: 1]
+
+      Module.register_attribute(__MODULE__, :telemetry_event,
+        accumulate: true,
+        persist: true
+      )
+
+      Module.register_attribute(__MODULE__, :telemetry_docs,
+        accumulate: true,
+        persist: true
+      )
+    end
+  end
+
+  @doc """
+  Declares a telemetry event. Accepts a telemetry event definition or telemetry event name. A full
+  definition is required for complete documentation and is highly encouraged.
+  """
+  defmacro telemetry_event(event) do
+    docstring =
+      case event do
+        event when is_list(event) ->
+          """
+          * `#{inspect(event)}`
+
+          """
+
+        {:%{}, _, event} ->
+          """
+          * `#{inspect(event[:event])}`
+            * Description: #{event[:description]}
+            * Measurements: `#{event[:measurements]}`
+            * Metadata: `#{event[:metadata]}`
+
+          """
+      end
+
+    quote do
+      @telemetry_event unquote(event)
+      @telemetry_docs unquote(docstring)
+    end
+  end
+
+  @doc """
+  Formats telemetry definitions declared in the module for inclusion in documentation.
+
+  Example
+
+  ```
+  ## Telemetry
+
+  #{format_telemetry_docs(@telemetry_docs)}
+  ```
+  """
+  @spec format_telemetry_docs([String.t()]) :: String.t()
+  def format_telemetry_docs(docs), do: Enum.reverse(docs) |> IO.iodata_to_binary()
```

Comment:
Besides the comment above about not storing `@telemetry_docs` upfront, I would make this a macro so we can hide the module attributes (i.e. they should be implementation details). Something like this:

```elixir
defmacro format_telemetry_docs do
  quote do
    TelemetryRegistry.__docs__(__MODULE__)
  end
end

@doc false
def __docs__(module) do
  Module.get_attribute(module, :telemetry_event)
  |> Enum.map(&format_event_to_doc/1)
  |> IO.iodata_to_binary()
end
```

WDYT?

Reply:
@josevalim module attributes cannot be implementation detail, as it would make Erlang implementation impossible from what I see. So I would say, that the module attributes will be **the** implementation.

Reply:
I am ok with with using `@telemetry_event` but the `@telemetry_docs` one should not be necessary/exposed.

FWIW, there are advantages to the macro approach, such as validating arguments, being able to provide defaults, etc. Given you need to `use TelemetryRegistry` module, I don't see the macro being a huge issue, BUT I can also see the appeal on keeping it close to the Erlang counter-part.

Reply:
I was copying from how ex_unit tests get rolled up. I prefer not to add a secondary one, but hadn't considered another macro to do it. I like it. 👍

Reply:
@josevalim, it still can be done in the function/macro that will generate the documentation text instead of there. IMHO that would make it clearer IMHO.

By the way, I was thinking about either adding the support for extensions to ExDoc or writing my own documentation generator that would use ASCIIDoc with support for extensions so it would be possible to generate documentation for other libraries that use metadata (like `OpenApiSpex`).

Reply:
>  it still can be done in the function/macro that will generate the documentation text instead of there.

Do you mean something like: `TelemetryRegistry.docs_for(SomeExternalModule)`? If so, I actually like this proposal too, as it decouples the definition from the docs.

> By the way, I was thinking about either adding the support for extensions to ExDoc 

Right - I don't doubt it is possible. The big question is when it is going to happen. Are we ok to postponing this feature indefinitely in favor of something that may or may not exist?

Reply:
For now I would go with moving documentation generation to `format_telemetry_docs/0` macro. So instead of using `telemetry_event` macro which generates default docs, generate them only when such macro will be used.

Reply:
Right, that's what i originally proposed. But now I think the latest API I proposed is better because it can be used to generate documentation for the current module and external modules.

It would be something like this:

    def docs_for(module) do
      get_events(module) |> format_docs()
    end

where 

    defp get_events(module) do
      Module.get_attribute(module, :telemetry_event)
    rescue
      _ -> module.__info__(:attributes)[:telemetry_event]
    end

Or something like that. So something like Phoenix can list `use TelemetryRegistry` and register the events on the modules that dispatch them but still have unified docs in a separate module.

Reply:
Why not:

```elixir
defp get_events(module) do
  {:docs_v1, _, _, _, _, metadata, _} = Code.fetch_docs(module)

  Map.get(metadata, :telemetry_events, [])
end
```

Which would also create base for future ExDoc integration.

Reply:
The above doesn't work if the module is being defined and we already store them in the attributes anyway, so I don't see what it would buy us. :)